src/app/clusters/soil-measurement-server/soil-measurement-server.cpp

```diff
@@ -0,0 +1,129 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "soil-measurement-server.h"
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <clusters/SoilMeasurement/Metadata.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::SoilMeasurement;
+using namespace chip::app::Clusters::SoilMeasurement::Attributes;
+
+static const Globals::Structs::MeasurementAccuracyRangeStruct::Type kDefaultSoilMoistureMeasurementLimitsAccuracyRange[] = {
+    { .rangeMin = 0, .rangeMax = 100, .percentMax = MakeOptional(static_cast<chip::Percent100ths>(10)) }
+};
+
+static const Globals::Structs::MeasurementAccuracyStruct::Type kDefaultSoilMoistureMeasurementLimits = {
+    .measurementType  = Globals::MeasurementTypeEnum::kSoilMoisture,
+    .measured         = true,
+    .minMeasuredValue = 0,
+    .maxMeasuredValue = 100,
+    .accuracyRanges   = DataModel::List<const Globals::Structs::MeasurementAccuracyRangeStruct::Type>(
+        kDefaultSoilMoistureMeasurementLimitsAccuracyRange)
+};
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace SoilMeasurement {
+
+Instance::Instance(EndpointId aEndpointId) :
+    AttributeAccessInterface(Optional<EndpointId>(aEndpointId), Id), mEndpointId(aEndpointId)
+{}
+
+Instance::~Instance()
+{
+    Shutdown();
+}
+
+CHIP_ERROR Instance::Init()
+{
+    VerifyOrDie(emberAfContainsServer(mEndpointId, Id) == true);
+
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+
+    // Initialize the soil moisture measurement limits to default values
+    mSoilMeasurementData.soilMoistureMeasurementLimits = kDefaultSoilMoistureMeasurementLimits;
```

Comment:
Applications have to use the setter to change this. Why not have it as a parameter in here and mostly needn't be changed unless the application wants to change it again at runtime?

Reply:
Indeed, this is somewhat a left over from the previous version of the implementation I did, so I agree this might as well be a parameter on init, I'm even wondering if the setter is really needed at that point, I guess it could be but on the other hand the application could also just create a new instance with the changes structure 🤔

Reply:
Changed in latest commit